includes/User.php

```diff
@@ -2082,41 +2082,48 @@ public static function clearUserCache( $userId ) {
 	 * wipes it, so the account cannot be logged in until
 	 * a new password is set, for instance via e-mail.
 	 *
-	 * @param $str String New password to set
-	 * @throws PasswordError on failure
+	 * @param string $password New password to set
+	 * @param bool   $forceLogout
 	 *
-	 * @return bool
+	 * @return bool on failure
+	 *
+	 * @throws PasswordError on failure
 	 */
-	public function setPassword( $str, $forceLogout = true ) {
-		global $wgAuth;
-
-		if( $str !== null ) {
-			if( !$wgAuth->allowPasswordChange() ) {
-				throw new PasswordError( wfMsg( 'password-change-forbidden' ) );
-			}
+	public function setPassword( $password, $forceLogout = true ) {
+		$heliosClient = self::heliosClient();
+		$heliosPasswordChange = null;
 
-			if( !$this->isValidPassword( $str ) ) {
-				global $wgMinimalPasswordLength;
-				$valid = $this->getPasswordValidity( $str );
-				if ( is_array( $valid ) ) {
-					$message = array_shift( $valid );
-					$params = $valid;
-				} else {
-					$message = $valid;
-					$params = array( $wgMinimalPasswordLength );
-				}
-				throw new PasswordError( wfMsgExt( $message, array( 'parsemag' ), $params ) );
-			}
+		try {
+			$heliosPasswordChange = $heliosClient->setPassword( $this->getId(), $password );
+		} catch ( \Exception $e ) {
+			WikiaLogger::instance()->error( 'Failed to communicate with Helios for password set', [
+				'userId' => $this->getId(),
+				'err'    => $e,
+			] );
 		}
 
-		if( !$wgAuth->setPassword( $this, $str ) ) {
+		if ( !$heliosPasswordChange ) {
```

Comment:
Not sure about placing this if here. In case of an exception, you will log the "Failed to communicate with Helios for password set" message twice - here and in the catch clause above. Also here you will log "Failed to communicate with Helios for password set" in case there response body was empty (as HeliosClientImpl::request returns null in that case). Maybe you should move this block inde the "try" block?